### DISCUSSION TITLE
[A-L] (8 maggio 2024) Esercizio 2

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#69] [Esercitazioni A-L]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/40034429/d2fcd2ee-4f66-4252-83bf-7cd49b6cde25)

**Comment by interestingmaneuver** [comment]
Se $r'(u, v) = 1 - r(u, v)$ calcoliamo i cammini minimi da $s$ (ad es. con Dijkstra), con il costo del cammino dato da $$c(p) = \Pi_{(u, v) \in p} \ r'(u, v)$$

La dimostrazione è semplice, $r'$ definisce la probabilità di fallimento per ogni arco. Siccome le probabilità sono indipendenti, l'intersezione è pari al prodotto.

**Comment by Elia-Belli** [comment]
## Dijkstra
Modifichiamo Dijkstra in modo da massimizzare $Dist[v] \cdot R[u]$ anziché minimizzare $Dist[v]+w(v,u)$, dove $Dist[v]$ è l'affidabilità del cammino più sicuro da $s\to v$ .
- **Complessità**: $O((n+m)\log n)$
```python
def cammini_più_sicuri(G,R,s):
    Dist = [0]*n
    P = [-1]*n
    P[s] = s
    H = [] #max_heap

    for v in G.V:  #setup heap
	H.insert(v,0)
    H.update(s,1)

    while(H):
	v = H.extract_max()
	Dist[v] = H.key(v)

	for u in v.adiacenti:
            if(u in H):
		alt = Dist[v] * R[u]
		if(alt > Dist[u]):
        	    Dist[u] = alt
		    H.update_key(u,alt)
		    P[u] = v
    return P
```

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/af1dcdee-8e46-42cd-a67c-0497cfcfa3bb)
per ritornare l'albero:
![immagine](https://github.com/user-attachments/assets/070e11e6-6b91-4938-a420-8684f732488b)
